test/e2e_node/container_lifecycle_test.go

```diff
@@ -3570,6 +3589,684 @@ var _ = SIGDescribe(nodefeature.SidecarContainers, "Containers Lifecycle", func(
 			time.Duration(containerTerminationSeconds+60)*time.Second)
 		framework.ExpectNoError(err, "the pod should be deleted before its terminationGracePeriodSeconds if the restartalbe init containers get termination signal correctly")
 	})
+
+	ginkgo.It("should mark pod as succeeded if any of the restartable init containers have terminated with non-zero exit code", func(ctx context.Context) {
+		restartableInit1 := "restartable-init-1"
+		restartableInit2 := "restartable-init-2"
+		restartableInit3 := "restartable-init-3"
+		regular1 := "regular-1"
+
+		podTerminationGracePeriodSeconds := int64(30)
+
+		pod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:       "restartable-init-terminated-with-non-zero-exit-code",
+				Finalizers: []string{testFinalizer},
+			},
+			Spec: v1.PodSpec{
+				RestartPolicy:                 v1.RestartPolicyNever,
+				TerminationGracePeriodSeconds: ptr.To(podTerminationGracePeriodSeconds),
+				InitContainers: []v1.Container{
+					{
+						Name:          restartableInit1,
+						Image:         busyboxImage,
+						RestartPolicy: &containerRestartPolicyAlways,
+						Command: ExecCommand(restartableInit1, execCommand{
+							Delay:              600,
+							TerminationSeconds: 1,
+							ExitCode:           0,
+						}),
+					},
+					{
+						Name:          restartableInit2,
+						Image:         busyboxImage,
+						RestartPolicy: &containerRestartPolicyAlways,
+						Command: ExecCommand(restartableInit2, execCommand{
+							Delay:              600,
+							TerminationSeconds: 1,
+							ExitCode:           1,
+						}),
+					},
+					{
+						Name:          restartableInit3,
+						Image:         busyboxImage,
+						RestartPolicy: &containerRestartPolicyAlways,
+						Command: ExecCommand(restartableInit3, execCommand{
+							Delay:              600,
+							TerminationSeconds: 1,
+							ExitCode:           0,
+						}),
+					},
+				},
+				Containers: []v1.Container{
+					{
+						Name:  regular1,
+						Image: busyboxImage,
+						Command: ExecCommand(regular1, execCommand{
+							Delay:              600,
+							TerminationSeconds: 1,
+							ExitCode:           0,
+						}),
+					},
+				},
+			},
+		}
+
+		preparePod(pod)
+
+		ginkgo.By("creating the pod with finalizer")
+		client := e2epod.NewPodClient(f)
+		pod = client.Create(ctx, pod)
+		defer client.RemoveFinalizer(ctx, pod.Name, testFinalizer)
```

Comment:
```suggestion
		ginkgo.DeferCleanup(client.RemoveFinalizer, pod.Name, testFinalizer)
```

Reply:
I tried that first, but the test failed, probably due to the impact of the DeferCleanup callstack. I'm not too familiar with Ginkgo, so I'm not entirely sure.


<details>
<summary><code>make test-e2e-node FOCUS="\[NodeFeature:SidecarContainers\] Containers Lifecycle" SKIP="\[Flaky\]|\[Slow\]|\[Disruptive\]|\[Serial\]"</code></summary>


```console
$ make test-e2e-node FOCUS="\[NodeFeature:SidecarContainers\] Containers Lifecycle" SKIP="\[Flaky\]|\[Slow\]|\[Disruptive\]|\[Serial\]"
(...)
  [FAILED] wait for pod "regular-prestop-exceeded-termination-grace-period" to disappear: expected pod to not be found: Timed out after 120.002s.
  Expected
      <*v1.Pod | 0x4001712d88>:
          metadata:
            annotations:
              owner.test: k8s.io/kubernetes/test/e2e_node/container_lifecycle_test.go:4094
            creationTimestamp: "2024-09-09T16:43:25Z"
            deletionGracePeriodSeconds: 0
            deletionTimestamp: "2024-09-09T16:43:31Z"
            finalizers:
            - example.com/test-finalizer
            managedFields:
            - apiVersion: v1
              fieldsType: FieldsV1
              fieldsV1:
                f:metadata:
                  f:annotations:
                    .: {}
                    f:owner.test: {}
                  f:finalizers:
                    .: {}
                    v:"example.com/test-finalizer": {}
                f:spec:
                  f:containers:
                    k:{"name":"regular-1"}:
                      .: {}
                      f:command: {}
                      f:image: {}
                      f:imagePullPolicy: {}
                      f:lifecycle:
                        .: {}
                        f:preStop:
                          .: {}
                          f:exec:
                            .: {}
                            f:command: {}
                      f:name: {}
                      f:resources:
                        .: {}
                        f:limits:
                          .: {}
                          f:memory: {}
                        f:requests:
                          .: {}
                          f:memory: {}
                      f:terminationMessagePath: {}
                      f:terminationMessagePolicy: {}
                      f:volumeMounts:
                        .: {}
                        k:{"mountPath":"/persistent"}:
                          .: {}
                          f:mountPath: {}
                          f:name: {}
                  f:dnsPolicy: {}
                  f:enableServiceLinks: {}
                  f:initContainers:
                    .: {}
                    k:{"name":"restartable-init-1"}:
                      .: {}
                      f:command: {}
                      f:image: {}
                      f:imagePullPolicy: {}
                      f:name: {}
                      f:resources:
                        .: {}
                        f:limits:
                          .: {}
                          f:memory: {}
                        f:requests:
                          .: {}
                          f:memory: {}
                      f:restartPolicy: {}
                      f:terminationMessagePath: {}
                      f:terminationMessagePolicy: {}
                      f:volumeMounts:
                        .: {}
                        k:{"mountPath":"/persistent"}:
                          .: {}
                          f:mountPath: {}
                          f:name: {}
                    k:{"name":"restartable-init-2"}:
                      .: {}
                      f:command: {}
                      f:image: {}
                      f:imagePullPolicy: {}
                      f:name: {}
                      f:resources:
                        .: {}
                        f:limits:
                          .: {}
                          f:memory: {}
                        f:requests:
                          .: {}
                          f:memory: {}
                      f:restartPolicy: {}
                      f:terminationMessagePath: {}
                      f:terminationMessagePolicy: {}
                      f:volumeMounts:
                        .: {}
                        k:{"mountPath":"/persistent"}:
                          .: {}
                          f:mountPath: {}
                          f:name: {}
                    k:{"name":"restartable-init-3"}:
                      .: {}
                      f:command: {}
                      f:image: {}
                      f:imagePullPolicy: {}
                      f:name: {}
                      f:resources:
                        .: {}
                        f:limits:
                          .: {}
                          f:memory: {}
                        f:requests:
                          .: {}
                          f:memory: {}
                      f:restartPolicy: {}
                      f:terminationMessagePath: {}
                      f:terminationMessagePolicy: {}
                      f:volumeMounts:
                        .: {}
                        k:{"mountPath":"/persistent"}:
                          .: {}
                          f:mountPath: {}
                          f:name: {}
                  f:nodeName: {}
                  f:restartPolicy: {}
                  f:schedulerName: {}
                  f:securityContext: {}
                  f:terminationGracePeriodSeconds: {}
                  f:volumes:
                    .: {}
                    k:{"name":"persistent"}:
                      .: {}
                      f:emptyDir: {}
                      f:name: {}
              manager: e2e_node.test
              operation: Update
              time: "2024-09-09T16:43:25Z"
            - apiVersion: v1
              fieldsType: FieldsV1
              fieldsV1:
                f:status:
                  f:conditions:
                    .: {}
                    k:{"type":"ContainersReady"}:
                      .: {}
                      f:lastProbeTime: {}
                      f:lastTransitionTime: {}
                      f:reason: {}
                      f:status: {}
                      f:type: {}
                    k:{"type":"Initialized"}:
                      .: {}
                      f:lastProbeTime: {}
                      f:lastTransitionTime: {}
                      f:status: {}
                      f:type: {}
                    k:{"type":"PodReadyToStartContainers"}:
                      .: {}
                      f:lastProbeTime: {}
                      f:lastTransitionTime: {}
                      f:status: {}
                      f:type: {}
                    k:{"type":"PodScheduled"}:
                      .: {}
                      f:lastProbeTime: {}
                      f:lastTransitionTime: {}
                      f:status: {}
                      f:type: {}
                    k:{"type":"Ready"}:
                      .: {}
                      f:lastProbeTime: {}
                      f:lastTransitionTime: {}
                      f:reason: {}
                      f:status: {}
                      f:type: {}
                  f:containerStatuses: {}
                  f:hostIP: {}
                  f:hostIPs: {}
                  f:initContainerStatuses: {}
                  f:phase: {}
                  f:podIP: {}
                  f:podIPs:
                    .: {}
                    k:{"ip":"10.10.9.7"}:
                      .: {}
                      f:ip: {}
                  f:startTime: {}
              manager: kubelet
              operation: Update
              subresource: status
              time: "2024-09-09T16:43:39Z"
            name: regular-prestop-exceeded-termination-grace-period
            namespace: containers-lifecycle-test-4025
            resourceVersion: "1384"
            uid: 612730ef-44db-4a45-876a-c830ab2353fc
          spec:
            containers:
            - command:
              - sh
              - -c
              - touch /persistent/regular-1.log; cat /persistent/regular-1.log >> /proc/1/fd/1;
                echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'regular-1 Starting 0' |
                tee -a /persistent/regular-1.log >> /proc/1/fd/1; _term() { sleep 20; echo `date
                -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'regular-1 Exiting' | tee -a /persistent/regular-1.log
                >> /proc/1/fd/1; exit 0; }; trap _term TERM; touch started; echo `date -u +%FT$(nmeter
                -d0 '%3t' | head -n1)Z` 'regular-1 Started' | tee -a /persistent/regular-1.log
                >> /proc/1/fd/1; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'regular-1
                Delaying 600' | tee -a /persistent/regular-1.log >> /proc/1/fd/1; exec sleep
                600 & wait $!; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'regular-1
                Exiting'  | tee -a /persistent/regular-1.log >> /proc/1/fd/1; exit 0
              image: registry.k8s.io/e2e-test-images/busybox:1.36.1-1
              imagePullPolicy: IfNotPresent
              lifecycle:
                preStop:
                  exec:
                    command:
                    - sh
                    - -c
                    - touch /persistent/regular-1.log; cat /persistent/regular-1.log >> /proc/1/fd/1;
                      echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'regular-1 Starting
                      0' | tee -a /persistent/regular-1.log >> /proc/1/fd/1; _term() { sleep
                      0; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'regular-1 Exiting'
                      | tee -a /persistent/regular-1.log >> /proc/1/fd/1; exit 0; }; trap _term
                      TERM; touch started; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z`
                      'regular-1 Started' | tee -a /persistent/regular-1.log >> /proc/1/fd/1;
                      echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'regular-1 Delaying
                      20' | tee -a /persistent/regular-1.log >> /proc/1/fd/1; exec sleep 20
                      & wait $!; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'regular-1
                      Exiting'  | tee -a /persistent/regular-1.log >> /proc/1/fd/1; exit 0
              name: regular-1
              resources:
                limits:
                  memory: 35Mi
                requests:
                  memory: 15Mi
              terminationMessagePath: /dev/termination-log
              terminationMessagePolicy: File
              volumeMounts:
              - mountPath: /persistent
                name: persistent
            dnsPolicy: Default
            enableServiceLinks: true
            initContainers:
            - command:
              - sh
              - -c
              - touch /persistent/restartable-init-1.log; cat /persistent/restartable-init-1.log
                >> /proc/1/fd/1; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'restartable-init-1
                Starting 0' | tee -a /persistent/restartable-init-1.log >> /proc/1/fd/1; _term()
                { sleep 1; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'restartable-init-1
                Exiting' | tee -a /persistent/restartable-init-1.log >> /proc/1/fd/1; exit 0;
                }; trap _term TERM; touch started; echo `date -u +%FT$(nmeter -d0 '%3t' | head
                -n1)Z` 'restartable-init-1 Started' | tee -a /persistent/restartable-init-1.log
                >> /proc/1/fd/1; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'restartable-init-1
                Delaying 600' | tee -a /persistent/restartable-init-1.log >> /proc/1/fd/1; exec
                sleep 600 & wait $!; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'restartable-init-1
                Exiting'  | tee -a /persistent/restartable-init-1.log >> /proc/1/fd/1; exit
                0
              image: registry.k8s.io/e2e-test-images/busybox:1.36.1-1
              imagePullPolicy: IfNotPresent
              name: restartable-init-1
              resources:
                limits:
                  memory: 35Mi
                requests:
                  memory: 15Mi
              restartPolicy: Always
              terminationMessagePath: /dev/termination-log
              terminationMessagePolicy: File
              volumeMounts:
              - mountPath: /persistent
                name: persistent
            - command:
              - sh
              - -c
              - touch /persistent/restartable-init-2.log; cat /persistent/restartable-init-2.log
                >> /proc/1/fd/1; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'restartable-init-2
                Starting 0' | tee -a /persistent/restartable-init-2.log >> /proc/1/fd/1; _term()
                { sleep 20; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'restartable-init-2
                Exiting' | tee -a /persistent/restartable-init-2.log >> /proc/1/fd/1; exit 0;
                }; trap _term TERM; touch started; echo `date -u +%FT$(nmeter -d0 '%3t' | head
                -n1)Z` 'restartable-init-2 Started' | tee -a /persistent/restartable-init-2.log
                >> /proc/1/fd/1; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'restartable-init-2
                Delaying 600' | tee -a /persistent/restartable-init-2.log >> /proc/1/fd/1; exec
                sleep 600 & wait $!; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'restartable-init-2
                Exiting'  | tee -a /persistent/restartable-init-2.log >> /proc/1/fd/1; exit
                0
              image: registry.k8s.io/e2e-test-images/busybox:1.36.1-1
              imagePullPolicy: IfNotPresent
              name: restartable-init-2
              resources:
                limits:
                  memory: 35Mi
                requests:
                  memory: 15Mi
              restartPolicy: Always
              terminationMessagePath: /dev/termination-log
              terminationMessagePolicy: File
              volumeMounts:
              - mountPath: /persistent
                name: persistent
            - command:
              - sh
              - -c
              - touch /persistent/restartable-init-3.log; cat /persistent/restartable-init-3.log
                >> /proc/1/fd/1; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'restartable-init-3
                Starting 0' | tee -a /persistent/restartable-init-3.log >> /proc/1/fd/1; _term()
                { sleep 1; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'restartable-init-3
                Exiting' | tee -a /persistent/restartable-init-3.log >> /proc/1/fd/1; exit 0;
                }; trap _term TERM; touch started; echo `date -u +%FT$(nmeter -d0 '%3t' | head
                -n1)Z` 'restartable-init-3 Started' | tee -a /persistent/restartable-init-3.log
                >> /proc/1/fd/1; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'restartable-init-3
                Delaying 600' | tee -a /persistent/restartable-init-3.log >> /proc/1/fd/1; exec
                sleep 600 & wait $!; echo `date -u +%FT$(nmeter -d0 '%3t' | head -n1)Z` 'restartable-init-3
                Exiting'  | tee -a /persistent/restartable-init-3.log >> /proc/1/fd/1; exit
                0
              image: registry.k8s.io/e2e-test-images/busybox:1.36.1-1
              imagePullPolicy: IfNotPresent
              name: restartable-init-3
              resources:
                limits:
                  memory: 35Mi
                requests:
                  memory: 15Mi
              restartPolicy: Always
              terminationMessagePath: /dev/termination-log
              terminationMessagePolicy: File
              volumeMounts:
              - mountPath: /persistent
                name: persistent
            nodeName: ubuntu
            preemptionPolicy: PreemptLowerPriority
            priority: 0
            restartPolicy: Never
            schedulerName: default-scheduler
            securityContext: {}
            terminationGracePeriodSeconds: 5
            tolerations:
            - effect: NoExecute
              key: node.kubernetes.io/not-ready
              operator: Exists
              tolerationSeconds: 300
            - effect: NoExecute
              key: node.kubernetes.io/unreachable
              operator: Exists
              tolerationSeconds: 300
            volumes:
            - emptyDir: {}
              name: persistent
          status:
            conditions:
            - lastProbeTime: null
              lastTransitionTime: "2024-09-09T16:43:39Z"
              status: "False"
              type: PodReadyToStartContainers
            - lastProbeTime: null
              lastTransitionTime: "2024-09-09T16:43:29Z"
              status: "True"
              type: Initialized
            - lastProbeTime: null
              lastTransitionTime: "2024-09-09T16:43:39Z"
              reason: PodFailed
              status: "False"
              type: Ready
            - lastProbeTime: null
              lastTransitionTime: "2024-09-09T16:43:39Z"
              reason: PodFailed
              status: "False"
              type: ContainersReady
            - lastProbeTime: null
              lastTransitionTime: "2024-09-09T16:43:25Z"
              status: "True"
              type: PodScheduled
            containerStatuses:
            - containerID: containerd://b707391db8fa9c3ebc7604bc294020f277bff3a8d41a98a98d352c8757d9e903
              image: registry.k8s.io/e2e-test-images/busybox:1.36.1-1
              imageID: registry.k8s.io/e2e-test-images/busybox@sha256:a9155b13325b2abef48e71de77bb8ac015412a566829f621d06bfae5c699b1b9
              lastState: {}
              name: regular-1
              ready: false
              restartCount: 0
              started: false
              state:
                terminated:
                  containerID: containerd://b707391db8fa9c3ebc7604bc294020f277bff3a8d41a98a98d352c8757d9e903
                  exitCode: 137
                  finishedAt: "2024-09-09T16:43:39Z"
                  reason: Error
                  startedAt: "2024-09-09T16:43:29Z"
              volumeMounts:
              - mountPath: /persistent
                name: persistent
            hostIP: 198.19.249.71
            hostIPs:
            - ip: 198.19.249.71
            initContainerStatuses:
            - containerID: containerd://2f915dbfcde0c6c9639fb3a7fd80914975ece3c4959a1c677a95a41b53a7aef3
              image: registry.k8s.io/e2e-test-images/busybox:1.36.1-1
              imageID: registry.k8s.io/e2e-test-images/busybox@sha256:a9155b13325b2abef48e71de77bb8ac015412a566829f621d06bfae5c699b1b9
              lastState: {}
              name: restartable-init-1
              ready: false
              restartCount: 0
              started: false
              state:
                terminated:
                  containerID: containerd://2f915dbfcde0c6c9639fb3a7fd80914975ece3c4959a1c677a95a41b53a7aef3
                  exitCode: 0
                  finishedAt: "2024-09-09T16:43:38Z"
                  reason: Completed
                  startedAt: "2024-09-09T16:43:26Z"
              volumeMounts:
              - mountPath: /persistent
                name: persistent
            - containerID: containerd://289dbc4ea5a33eafca903b4ea2d7002e9abe7afde7afb8e00d88908826642e1c
              image: registry.k8s.io/e2e-test-images/busybox:1.36.1-1
              imageID: registry.k8s.io/e2e-test-images/busybox@sha256:a9155b13325b2abef48e71de77bb8ac015412a566829f621d06bfae5c699b1b9
              lastState: {}
              name: restartable-init-2
              ready: false
              restartCount: 0
              started: false
              state:
                terminated:
                  containerID: containerd://289dbc4ea5a33eafca903b4ea2d7002e9abe7afde7afb8e00d88908826642e1c
                  exitCode: 137
                  finishedAt: "2024-09-09T16:43:39Z"
                  reason: Error
                  startedAt: "2024-09-09T16:43:27Z"
              volumeMounts:
              - mountPath: /persistent
                name: persistent
            - containerID: containerd://7d5687ee1693b4b085bc44546ec77caae534a381bbecb81c574647a25cc657d8
              image: registry.k8s.io/e2e-test-images/busybox:1.36.1-1
              imageID: registry.k8s.io/e2e-test-images/busybox@sha256:a9155b13325b2abef48e71de77bb8ac015412a566829f621d06bfae5c699b1b9
              lastState: {}
              name: restartable-init-3
              ready: false
              restartCount: 0
              started: false
              state:
                terminated:
                  containerID: containerd://7d5687ee1693b4b085bc44546ec77caae534a381bbecb81c574647a25cc657d8
                  exitCode: 0
                  finishedAt: "2024-09-09T16:43:38Z"
                  reason: Completed
                  startedAt: "2024-09-09T16:43:28Z"
              volumeMounts:
              - mountPath: /persistent
                name: persistent
            phase: Failed
            podIP: 10.10.9.7
            podIPs:
            - ip: 10.10.9.7
            qosClass: Burstable
            startTime: "2024-09-09T16:43:25Z"
  to be nil
  In [AfterEach] at: k8s.io/kubernetes/test/e2e/framework/pod/pod_client.go:224 @ 09/10/24 01:45:39.812
(...)

Summarizing 6 Failures:
  [FAIL] [sig-node] [NodeFeature:SidecarContainers] Containers Lifecycle [AfterEach] should mark pod as succeeded if any of the prestop hook in restartable init containers have exceeded its termination grace period seconds [sig-node, NodeFeature:SidecarContainers]
  k8s.io/kubernetes/test/e2e/framework/pod/pod_client.go:224
  [FAIL] [sig-node] [NodeFeature:SidecarContainers] Containers Lifecycle [AfterEach] should mark pod as succeeded if any of the restartable init containers have exceeded its termination grace period seconds [sig-node, NodeFeature:SidecarContainers]
  k8s.io/kubernetes/test/e2e/framework/pod/pod_client.go:224
  [FAIL] [sig-node] [NodeFeature:SidecarContainers] Containers Lifecycle [AfterEach] should mark pod as succeeded if any of the restartable init containers have terminated with non-zero exit code [sig-node, NodeFeature:SidecarContainers]
  k8s.io/kubernetes/test/e2e/framework/pod/pod_client.go:224
  [FAIL] [sig-node] [NodeFeature:SidecarContainers] Containers Lifecycle [AfterEach] should mark pod as failed if regular container has exceeded its termination grace period seconds [sig-node, NodeFeature:SidecarContainers]
  k8s.io/kubernetes/test/e2e/framework/pod/pod_client.go:224
  [FAIL] [sig-node] [NodeFeature:SidecarContainers] Containers Lifecycle [AfterEach] should mark pod as failed if any of the prestop hook in regular container has exceeded its termination grace period seconds [sig-node, NodeFeature:SidecarContainers]
  k8s.io/kubernetes/test/e2e/framework/pod/pod_client.go:224
  [FAIL] [sig-node] [NodeFeature:SidecarContainers] Containers Lifecycle [AfterEach] should mark pod as succeeded if any of the restartable init containers have terminated with non-zero exit code by prestop hook [sig-node, NodeFeature:SidecarContainers]
  k8s.io/kubernetes/test/e2e/framework/pod/pod_client.go:224

Ran 82 of 661 Specs in 303.128 seconds
FAIL! -- 76 Passed | 6 Failed | 0 Pending | 579 Skipped
```

</details>